File: projects/m5/002-polygon-area-calculator/js/shape_calculator.js
class Rectangle {
  #name = "Rectangle";

  constructor(width, height) {
    if (width < 0 || height < 0) {
      console.log(
        "Error!: The input must be a positive number\nthe height and width value has been set to 0"
      );
      this._width = 0;
      this._height = 0;
    } else {
      this._width = width;
      this._height = height;
    }
  }

  getWidth() {
    return this._width;
  }

  setWidth(value) {
    if (value < 0) {
      console.log(
        `Error!: The input must be a positive number\nthe width has been set to the previous value (${this._width})`
      );
      value = this._width;
    }
    this._width = value;
  }

  getHeight() {
    return this._height;
  }

  setHeight(value) {
    if (value < 0) {
      console.log(
        `Error!: The input must be a positive number\nthe height has been set to the previous value (${this._height})`
      );
      value = this._height;
    }
    this._height = value;
  }

  getArea() {
    return this._width * this._height;
  }

  getPerimeter() {
    return 2 * this._width + 2 * this._height;
  }

  getDiagonal() {
    return Math.pow(Math.pow(this._width, 2) + Math.pow(this._height, 2), 0.5);
  }

  checkSide(side) {
    return side > 50 ? true : false;
  }

  getPicture() {
    if (this.checkSide(this._width) || this.checkSide(this._height)) {
      return "Too big for picture.";
    }

    let picture = "";

    for (let y = 0; y < this._height; y++) {
      for (let x = 0; x < this._width; x++) {
        picture += "*";
      }
      picture += "\n";
    }
    return picture;
  }

  getString() {
    return `${this.#name}(width=${this._width}, height=${this._height})`;
  }

  getAmountInside(shape) {
    return shape.getArea() > this.getArea()
      ? 0
      : Math.floor(this.getArea() / shape.getArea());
  }
}

//Class Square
class Square extends Rectangle {
  #name = "Square";

  constructor(side) {
    super(side, side);
  }

  setSide(value) {
    if (value < 0) {
      console.log(
        `Error!: The Input must be a positive number\nwidth and height have been set to their previous values (${this._width})`
      );
      value = this._width;
    }
    this._width = value;
    this._height = value;
  }

  getString() {
    return `${this.#name}(side=${this._width})`;
  }
}

module.exports = {Rectangle, Square};

File: projects/m5/002-polygon-area-calculator/solutions/federicosiddi/shape_calculator.py
class Rectangle:
    def __init__(self, width, height):
        self.width = width
        self.height = height
    
    def __str__(self):
        output = f"Rectangle(width={self.width}, height={self.height})"
        return output
    
    def set_width(self, width):
        self.width = width
    
    def set_height(self, height):
        self.height = height
    
    def get_area(self):
        return self.width * self.height
    
    def get_perimeter(self):
        return 2 * self.width + 2 * self.height
    
    def get_diagonal(self):
        return (self.width ** 2 + self.height ** 2) ** .5
    
    def get_picture(self):
        if self.height > 50 or self.width > 50:
            return "Too big for picture."
        else:
            return (("*" * self.width) + "\n") * self.height
    
    def get_amount_inside(self, shape):
        res = (self.width // shape.width) * (self.height // shape.height)
        return res


class Square(Rectangle):
    def __init__(self, side):
        self.width = side
        self.height = side
    
    def __str__(self):
        output = f"Square(side={self.width})"
        return output

    def set_side(self, side):
        self.width = side
        self.height = side
    
    def set_width(self, side):
        self.set_side(side)
    
    def set_height(self, side):
        self.set_side(side)

# TEST
if __name__ == "__main__":
    rect = Rectangle(5, 10)
    print(rect.get_area())
    rect.set_width(3)
    print(rect.get_perimeter())
    print(rect)

    sq = Square(9)
    print(sq.get_area())
    sq.set_side(4)
    print(sq.get_diagonal())
    print(sq)